### PR TITLE
Fix naive filtering of uploads directory.

### LINKED ISSUE #58696: Font Library: "Fetch error: The response is not a valid JSON response" when filtering font_dir.
### Description

When testing a filter to change the directory of downloaded fonts, the fonts are not installed and return an error instead. Using the example in #57697 as a base.

Example code:
```php
/**
 * Define a custom font directory for the WP Font Library.
 */
function alter_wp_fonts_dir( $defaults ) {
	$wp_upload_dir = wp_get_upload_dir();
	$uploads_basedir = $wp_upload_dir['basedir'];
	$uploads_baseurl = $wp_upload_dir['baseurl'];

	$fonts_dir = $uploads_basedir . '/fonts';
	// Generate the URL for the fonts directory from the font dir.
	$fonts_url = str_replace( $uploads_basedir, $uploads_baseurl, $fonts_dir );

	$defaults['path'] = $fonts_dir;
	$defaults['url']  = $fonts_url;

	return $defaults;
}
add_filter( 'font_dir', 'alter_wp_fonts_dir' );
```

### Step-by-step reproduction instructions

1. Hook the code above to a function bootstrapped on `init`.
2. Navigate to the Font Library
3. Navigate to Install Fonts
4. Pick a font to install
5. Click the Install button

### Screenshots, screen recording, code snippet

[Loom recording](https://www.loom.com/share/29a26c5a3060422a9598c8f6a2204eb6?sid=f5e8b2d0-adfc-457a-b200-e03fddcdf481)

Console:
Screenshot 2024-02-05 at 3 08 10 PM

### Environment info

- WordPress 6.5-alpha-57531
- PHP: tested on 7.4.33, 8.1.18, 8.2.5
- Gutenberg: tested on 17.6.1, 17.6.2, 17.6.4
- MacOS Sonoma 14.2.1
- Google Chrome 121.0.6167.85

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by peterwilsoncc**
Prevent infinite loops getting the font directory.

Fast follow for r57740 / 9c415ed5fb9b4069208b9e43ff774acbfb3ef3a6

This allows for WordPress extenders to filter the uploads directory to the font directory in the typical WordPress fashion, ie `add_filter( 'upload_dir', 'wp_get_font_dir' );`.

On it's own this would work but can trigger infinite loops on hosts calling `wp_upload_dir()` or `wp_get_upload_dir` on the `font_dir` filter.

It looks like this will need to be committed to wordpress-develop prior to Gutenberg.

Fixes https://github.com/WordPress/gutenberg/issues/58696
Fixes https://core.trac.wordpress.org/ticket/60652

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by peterwilsoncc** [comment]
@matiasbenedetto Ping. Unable to do so by adding you as a reviewer, sorry. I really need to fix that.

File: tests/phpunit/tests/fonts/font-library/wpFontsDir.php
Comment (mukeshpanchal27) [inline tests/phpunit/tests/fonts/font-library/wpFontsDir.php:87]:
```suggestion
			static function ( $upload_dir ) {
```

**Comment by swissspidy** [comment]
What I did not like about this prior `wp_get_font_dir` design is that it‘s actually quite confusing. The name suggests you can just use this function to get the font upload dir. But that‘s not true at all. It‘s only used to temporarily filter the upload location.

So if we want to keep the filter like this, we should really rename it to something like `_wp_filter_fonts_upload_dir()` or something

**Comment by peterwilsoncc** [comment]
> So if we want to keep the filter like this, we should really rename it to something like `_wp_filter_fonts_upload_dir()` or something

I agree. I had something to convert `wp_get_font_dir()` to a getter but decided to take the KISS approach. I'll work on renaming it and figuring out if it needs to go in the GB or wp-dev repo first.

**Comment by peterwilsoncc** [comment]
@swissspidy The function is called directly and used as a getter during the deletion of posts. I'll look at splitting it out so that WordPress runs a filter on `font_dir` or something similar.

https://github.com/WordPress/wordpress-develop/blob/b64bfae6ef2e0d06225ae79af4629598b1bbd6b6/src/wp-includes/fonts.php#L191

**Comment by matiasbenedetto** [comment]
> What I did not like about this prior wp_get_font_dir design is that it‘s actually quite confusing. The name suggests you can just use this function to get the font upload dir. But that‘s not true at all.

@swissspidy I don't get this. wp_get_font_dir() returns the font upload dir. What do you mean that this function can't be used to get the font upload dir?

**Comment by costdev** [comment]
I'm not entirely sure that this is needed when [this PR](https://github.com/WordPress/gutenberg/pull/58839) was merged in Gutenberg. @peterwilsoncc Can you explain why this is needed aswell?

**Comment by matiasbenedetto** [comment]
@costdev I think the rationale is here: https://github.com/WordPress/gutenberg/issues/58696#issuecomment-1974930217

**Comment by matiasbenedetto** [comment]
This is testing fine on my end. I tried these cases:

1. without adding any additional filter.
2. Removing the filter with:
```php
 add_filter( 'upload_dir', function ( $ul ) {
	remove_filter( 'upload_dir', 'wp_get_font_dir' );
	return $ul;
}, 5 );
```
3. Altering the filter with:
```php
/**
 * Define a custom font directory for the WP Font Library.
 */
function alter_wp_fonts_dir( $defaults ) {
	$wp_upload_dir = wp_get_upload_dir();
	$uploads_basedir = $wp_upload_dir['basedir'];
	$uploads_baseurl = $wp_upload_dir['baseurl'];

	$fonts_dir = $uploads_basedir . '/fonts';
	// Generate the URL for the fonts directory from the font dir.
	$fonts_url = str_replace( $uploads_basedir, $uploads_baseurl, $fonts_dir );

	$defaults['path'] = $fonts_dir;
	$defaults['url']  = $fonts_url;

	return $defaults;
}
add_filter( 'font_dir', 'alter_wp_fonts_dir' );
```

In the 3 cases, the fonts were successfully written in the expected folder:
1. /wp-content/fonts
2. /wp-content/uploads/`<year>`/`<month>`/
3. /wp-content/uploads/fonts

---

Apart from that, I agree about that receiving a parameter in `wp_get_font_dir( $font_dir = array() )`, seems contradictory with its name. It would be nice to split it into two functions as proposed here.

**Comment by peterwilsoncc** [comment]
I've pushed 19cc1c71176e5f351ee8feca0e103426b2bc2904 which is **nothing more than a thought bubble at this stage.**

`wp_get_font_dir()` and `wp_font_dir()` are to mirror the names and APIs of the `*_upload_dir()` functions without the time component.

`wp_font_dir()` includes a call to `wp_upload_dir()` to allow removing the filter on `upload_dir, 5` when using `wp_get_font_dir()` as a getter in `_wp_before_delete_font_face()`. IE this is for consistent behavior with:

```php
 add_filter( 'upload_dir', function ( $ul ) {
	remove_filter( 'upload_dir', 'wp_get_font_dir' );
	return $ul;
}, 5 );
```

I don't really like the need for extending the number of functions but figured I'd push up the thought bubble so I can get feedback on it and figure out how to take a cleaner approach that allows for consistency using the new functions as a getter.

**Comment by peterwilsoncc** [comment]
Folks, I think this is ready for a review now. TBH, I'm not sure we need `wp_get_font...` and `wp_font...` to mirror their upload counterparts but I figured I'd push it for now.

It will obviously need some follow up work to allow for the fallback but it would be good to get the approach validated in parallel to https://github.com/WordPress/gutenberg/issues/59699 

cc @youknowriad @matiasbenedetto @swissspidy @dream-encode

File: tests/phpunit/tests/fonts/font-library/wpFontsDir.php
Comment (peterwilsoncc) [reply tests/phpunit/tests/fonts/font-library/wpFontsDir.php:87]: The function uses the `$this` object so can't be a static function. This the commit and revert.

**Review by mikachan — COMMENTED** [review]
👋 I've tested using similar steps to @matiasbenedetto's [here](https://github.com/WordPress/wordpress-develop/pull/6211#issuecomment-1979516854):

- Without adding any additional filter, fonts were uploaded to `/wp-content/fonts` ✅
- Removing the filter with `remove_filter( 'upload_dir', 'wp_apply_font_dir_filter' );`, fonts were uploaded to `/wp-content/uploads/<year>/<month>/` ✅ 
- However, when removing the filter with `remove_filter( 'upload_dir', 'wp_get_font_dir' );`, this didn't work, is this correct?
- Altering the filter with the same `alter_wp_fonts_dir` in the above comment, fonts were uploaded to `/wp-content/uploads/fonts` ✅

I understand the hesitance in extending the number of functions but I think mirroring the `*_upload_dir()` functions will make it easier for extenders to understand, so I like this approach. Not strongly opinionated here though if anyone has any other thoughts.

**Comment by matiasbenedetto** [comment]
> to mirror their upload counterparts

I'm not sure if replacing a function with three new ones is the best approach. I looked into the core's codebase and couldn't find any function similar to the proposed `wp_apply_font_dir_filter`. 

```php
function wp_apply_font_dir_filter( $font_dir ) {
	if ( doing_filter( 'font_dir' ) ) {
		// Avoid an infinite loop.
		return $font_dir;
	}

	return apply_filters( 'font_dir', $font_dir );
}
```

That's probably because no other function has this type of guard we are trying to add in this PR. Should this kind of prevention be implemented for all the functions that could filter 'upload_dir'? If the answer is no, why should we add it to the font_dir one?

**Comment by peterwilsoncc** [comment]
> Should this kind of prevention be implemented for all the functions that could filter 'upload_dir'? If the answer is no, why should we add it to the font_dir one?

Yes, it will need to be.

The current code in trunk fires the `upload_dir` filter when saving fonts in the REST API endpoint, but does not fire the filter when deleting font files via the hook in `wp_delete_post()`. 

Firing the hooks consistently is more important now that we've been asked to fall back to the uploads directory. Offloaders filter the upload directory to achieve this. Eg [WP VIP](https://github.com/Automattic/vip-go-mu-plugins/blob/dd73b8204e06fe32d6fc68732906ada0c4b69abd/files/class-vip-filesystem.php#L96) and [S3 Uploads](https://github.com/humanmade/S3-Uploads/blob/9ce35d7cf932d4f247a3cb96c8d6f26dd4286885/inc/class-plugin.php#L106).

> I looked into the core's codebase and couldn't find any function similar to the proposed `wp_apply_font_dir_filter`.

WordPress does have an existing pattern for functions wrapping the firing of hooks, for example [`wp_enqueue_scripts()`](https://github.com/WordPress/wordpress-develop/blob/2e60defa5d9bac1a3900f1ed5d92f56ce00f5d47/src/wp-includes/script-loader.php#L2256-L2263). I don't love the name I've chosen (suggestions are welcome) but I didn't want to use `font_dir` as it could lead to confusion with the getter functions.

File: src/wp-includes/fonts.php
Comment (peterwilsoncc) [inline src/wp-includes/fonts.php]: With the need to fallback to the uploads directory this may need to run later (say priority 20) to allow for plugins that filter the uploads directory to an offloader at the default priority, 10. This will avoid an order of operations issue as hooks running at the same priority run in the order they are added.

File: src/wp-includes/fonts.php
Comment (costdev) [inline src/wp-includes/fonts.php]:
```suggestion
 * Applies filters to manipulate the font directory used by the font library.
 * 
 * Wrapper for apply_filters( 'font_dir', $font_dir ).
 * Does not apply filters if already doing the {@see 'font_dir'} filter.
```

File: src/wp-includes/fonts.php
Comment (costdev) [inline src/wp-includes/fonts.php]:
```suggestion
 * Modifies the default font directory.
```

File: tests/phpunit/tests/fonts/font-library/wpFontsDir.php
Comment (costdev) [inline tests/phpunit/tests/fonts/font-library/wpFontsDir.php]: Rather than perform an assertion each time, what about this?
```php
if ( $count >= 5 ) {
	$this->fail( 'Filtering the uploads directory triggered an infinite loop.' );
}
```

Comment (peterwilsoncc) [reply tests/phpunit/tests/fonts/font-library/wpFontsDir.php]: Done in 5117eef291bd58a9bd1eb1c104aaf0c77b0e91c3

File: src/wp-includes/fonts.php
Comment (matiasbenedetto) [inline src/wp-includes/fonts.php]: Could this function be called `wp_default_font_dir` instead? The `_filter` part seems unnecessary, and I haven't seen any other core function called in that fashion.

Comment (peterwilsoncc) [reply src/wp-includes/fonts.php]: Done in 1c870cc5dd1b4a97253eb062f0127f6bcc5c8050

File: tests/phpunit/tests/fonts/font-library/wpFontsDir.php
Comment (costdev) [inline tests/phpunit/tests/fonts/font-library/wpFontsDir.php]:
```suggestion
		 * This emulates moving font files back to the uploads directory due
```

File: src/wp-includes/fonts.php
Comment (costdev) [inline src/wp-includes/fonts.php]: I've been thinking about the name of this function. While still not sure on the general name, I'd suggest that we at least call it `wp_apply_font_dir_filters()` to reflect the docblock, and the fact it applies _all_ `'font_dir'`filters.

Comment (peterwilsoncc) [reply src/wp-includes/fonts.php]: Done in https://github.com/WordPress/wordpress-develop/pull/6211/commits/52103f63b6905f3bcb980303b3bfa9905a58c9d2

**Comment by azaozz** [comment]
> This allows for WordPress extenders **to filter the uploads directory to the font directory** in the typical WordPress fashion, ie add_filter( 'upload_dir', 'wp_get_font_dir' );.

I'm a bit unsure why would someone want to do this (emphasis added)? Uploads dir is the uploads dir, and the fonts dir is... the fonts dir? Same for the plugins dir, themes dir, etc.?

Yea, the code that implements uploading of fonts reuses `wp_handle_upload()`, and there is a lambda, non-removable filter to make that work (which doesn't seem needed). 

May be missing something but it seems it may be a good idea to just remove that `add_filter( 'upload_dir', $set_upload_dir )` filter from `class-wp-rest-font-faces-controller.php` and pass the actual `$fonts_dir` data as another override to `wp_handle_upload()`? This should remove any possibilities for loops, etc.

**Comment by azaozz** [comment]
Sorry, didn't mean to close :(

**Comment by peterwilsoncc** [comment]
> May be missing something but it seems it may be a good idea to just remove that `add_filter( 'upload_dir', $set_upload_dir )` filter from `class-wp-rest-font-faces-controller.php` and pass the actual `$fonts_dir` data as another override to `wp_handle_upload()`? This should remove any possibilities for loops, etc.

@azaozz There is no way to pass a custom directory as an override. The upload handlers simply call `wp_upload_dir()`

https://github.com/WordPress/wordpress-develop/blob/14d55b80dc99e917a0e6ac266b8069cd36018e03/src/wp-admin/includes/file.php#L975-L982

A use case for an extender may be to make use of the feature outside of the site-editor. For example via the customizer for a classic theme or elsewhere for a hybrid theme. I'm not sure what else extenders will come up with but I'm often impressed by their ingenuity when extending WP.

My main concern with the lambda filter is that it works around the bug rather than fixes the bug. I've worked a lot on the feature over the last few weeks and don't trust myself to remember that I need a work-around if adding code elsewhere, let alone another contributor who has been less involved.

At a defensive coding level, if it's just as simple to fix a bug as to work around it then Core should always go with the former.

**Comment by azaozz** [comment]
> The upload handlers simply call wp_upload_dir()

Right. So the "override" could probably be something as simple as:
``` php
if ( isset( $overrides['font_upload'] ) ) {
    $uploads = wp_get_font_dir( 'create' ); // Passing "create" to indicate to make the dir if needed
} else {
    $uploads = wp_upload_dir( $time );
}

/*
 * A writable uploads dir will pass this test. Again, there's no point
 * overriding this one.
 */
if ( ! ( $uploads && false === $uploads['error'] ) ) {
    return call_user_func_array( $upload_error_handler, array( &$file, $uploads['error'] ) );
}
```

Internally wp_get_font_dir() uses wp_upload_dir() if /fonts has to be moved there. Both of these functions have filters, etc. so extenders will have no problems hooking/tweaking anything they need?

(I know, the "design" with overrides here is super old and non-intuitive (imho). It's been there since WP 1.5 I think.)

**Comment by peterwilsoncc** [comment]
I think we'd need to reproduce `wp_upload_dir()` and some of `_wp_upload_dir()` rather than use the filter.

Mainly, though, it's as you say: it's extending an unintuitive API. It's also pretty easy to accidentally remove some of the security built in to the handlers using it so encouraging it's use further doesn't seem like a good idea. e3a520606dcfec7aec65b8fe94822183882f3055 is an example of where it can lead to problems.

**Comment by creativecoder** [comment]
I've been working through this PR over the last couple of days, and here are my notes:

Given the constraints of

- Not modifying `wp_handle_upload`
- Wanting to directly apply `wp_font_dir` to the `upload_dir` filter
- While also using `wp_get_upload_dir()` in a function applied to the `font_dir` filter

I can't find a better way of handling this than what we have in this PR.

---

That said, stepping back, the implementation does still seem a bit odd to me, and I wonder if that's a hint there's a better way. For example, the default font path is created `add_filter( 'font_dir', 'wp_default_font_dir', 5 );`. If that default filter is removed and no others are added, the font directory becomes to the (filtered) value `wp_upload_dir`, which doesn't seem right.

That leads me to wondering, what other filters are out there in the wild for `upload_dir`? What are the chances that those will override the font-faces endpoint filtering of upload_dir and put fonts in a location intended for uploads? Perhaps, at the least, the `upload_dir` filter in the font-faces endpoint should be running on a later priority. Or maybe we should modify `wp_handle_upload` to more directly account for uploading fonts vs. uploads, as @azaozz suggested.

---

Also, if the most common use case for changing the fonts directory is moving it to a different, singular location (e.g. `/uploads/fonts`), but not doing anything dynamic, perhaps the addition of a constant and/or a canonical plugin would handle a lot of the concern more simply. Those seem like they would be easier to implement for the majority of cases, and the complexity of using the `font_dir` filter as is seems like less of a problem.

**Comment by peterwilsoncc** [comment]
> That said, stepping back, the implementation does still seem a bit odd to me, and I wonder if that's a hint there's a better way. For example, the default font path is created `add_filter( 'font_dir', 'wp_default_font_dir', 5 );`. If that default filter is removed and no others are added, the font directory becomes to the (filtered) value `wp_upload_dir`, which doesn't seem right.

That's fair. I can rip that out if you wish and merge it in to a renamed `wp_apply_font_dir_filters`.

A part of my thinking was that it would make it easier for extenders and hosts to switch to `/uploads/fonts` but I see your point -- it makes it a little to easy to switch to `uploads/2024/03/font.woff`

**Comment by peterwilsoncc** [comment]
> That leads me to wondering, what other filters are out there in the wild for `upload_dir`? What are the chances that those will override the font-faces endpoint filtering of upload_dir and put fonts in a location intended for uploads? Perhaps, at the least, the `upload_dir` filter in the font-faces endpoint should be running on a later priority. Or maybe we should modify `wp_handle_upload` to more directly account for uploading fonts vs. uploads, as @azaozz suggested.

I did a search of the plugin repo for [`["']upload_dir['"]`](https://wpdirectory.net/search/01HSFMXQEBG4YFTRR6BZN5XSBX) and couldn't find much in the way of late filtering of the uploads directory that will concern us -- most late filtering appeared to be targeting a particular plugins uploads rather than all uploads.

I'm still not keen on adding another override to handling uploads. Even accepting that `$overrides['font_upload']` works (and it likely would) I think it's premature to start adding context options to upload directories and upload handlers until further information is received about future first class objects.

I don't want to introduce a specific `font_upload` override in the final week of the release cycle, only to learn in a few releases that there's a need for another context and we've locked WordPress in to a pattern focused on a single feature.

**Comment by youknowriad** [comment]
I've been looking at this PR a bit and I'm not entirely sure I follow everything just yet. 

> This allows for WordPress extenders to filter the uploads directory to the font directory in the typical WordPress fashion, ie add_filter( 'upload_dir', 'wp_get_font_dir' );.

Personally, I agree with @azaozz I don't know why anyone would do this and if I'm reading properly, it's because there's no way to indicate the target directory properly in the wp upload function. So,I think I'd favor improving the upload primitive instead. 

But improving the upload primitive, whether that means a newer function or modifications to the existing function or something else feels like a bigger undertaking at this point.

So I'm actually fine with this PR.

**Comment by swissspidy** [comment]
> So,I think I'd favor improving the upload primitive instead. But improving the upload primitive, whether that means a newer function or modifications to the existing function or something else feels like a bigger undertaking at this point.

That's also my key takeaway here. Given that there will be likely more folders like this in the future, we should definitely overhaul this whole thing in the future. That also means any modifications now should keep that in mind.

So the approach here to fix the infinite loop seems fine to me as a start.

Just a couple of suggestions/thoughts on the code itself:

* `function wp_get_font_dir() { return wp_font_dir( false ); }` is a bit pointless. I don't see the need for this extra wrapping function. Can we just remove this, and rename `wp_font_dir` back to `wp_get_font_dir` like it was?
* Maybe prefix `wp_filter_font_directory` with an underscore for consistency with other "private" functions.

**Comment by azaozz** [comment]
> Given the constraints of
>    Not modifying wp_handle_upload
>    ...
> I can't find a better way of handling this than what we have in this PR.

Yep, I agree. The current code implements the needed functionality despite the constrains. My guess is that these constrains are left over from when that functionality was implemented in a plugin. (There is a better way to do that btw. Instead of trying to solve similar cases by imposing constrains, some temporary code can be introduced while WP is in development, and removed before release. All "feature plugins" can request such temp code, filters, settings, constants, etc.)

The question is: why not modify wp_handle_upload() in a simple, easy to follow, backwards compatible way? Imho the current patch here has higher chances to introduce errors/edge cases, and, as it uses nested filters, possible interference from existing code.

The fact is that wp_handle_upload() is a function that is re-used for new functionality. It is a standard practice to extend the functionality of the existing code to accommodate the new requirements. 

Yes, it is usually possible to use filters to accomplish the same tasks. However this doesn't make sense unless the filters are designed to also be used by plugin and theme authors. I.e. the filters that are used to extend the existing WP code and implement the new functionality would need to "make sense" and be useful for plugins. I somehow don't think this is the case here.

**Comment by azaozz** [comment]
> But improving the upload primitive, whether that means a newer function or modifications to the existing function or something else feels like a bigger undertaking at this point.

Hmm, I tend to disagree. Adding support for another (simple) parameter to a function seems quite "safer" as there is no chance to break backwards compatibility or to introduce edge cases to existing code. 

On the other hand using nested filters is a lot more complex, has some chance to introduce both back-compat problems and edge cases with existing plugins or themes.

> So I'm actually fine with this PR.

Thinking that if this patch is to be added to WP 6.5 few days before release at least the new functions have to be private, hidden, and not accessible by plugins in any way. That would allow this code to be improved and these functions to be removed much easier, instead of having to support them "forever". 

> Given that there will be likely more folders like this in the future, we should definitely overhaul this whole thing in the future.

Big +1!

To account for this thinking that this code needs to be "fixed" and any helper functions made private and added to a locked-down class, etc.

**Comment by peterwilsoncc** [comment]
> Thinking that if this patch is to be added to WP 6.5 few days before release at least the new functions have to be private, hidden, and not accessible by plugins in any way. That would allow this code to be improved and these functions to be removed much easier, instead of having to support them "forever".

The new functions need to be accessible to WordPress so are required to be public.

> I agree with @azaozz I don't know why anyone would do this and if I'm reading properly, it's because there's no way to indicate the target directory properly in the wp upload function. So,I think I'd favor improving the upload primitive instead.

I'm not sure what extenders will do either. A part of my concern is for future WordPress contributors unaware of the history. As the FL gets extended we need to avoid the infinite loop trap as a defensive coding technique.

> * `function wp_get_font_dir() { return wp_font_dir( false ); }` is a bit pointless. I don't see the need for this extra wrapping function. Can we just remove this, and rename `wp_font_dir` back to `wp_get_font_dir` like it was?

This mirrors the uploads API,  `wp_upload_dir()` and `wp_get_upload_dir`, one creates the directory and the other does not. Having `$create_dir` default to true in one API and false in another seems quite unintuitive.

**Comment by peterwilsoncc** [comment]
> Maybe prefix wp_filter_font_directory with an underscore for consistency with other "private" functions.

Done in 22764e3650e38b56737f5f558d9a2125710b08dc

In cf211ae8117adbdc844ddd70bef04301c01c094f I added a code sample for using the filter and have confirmed the docs-parser correctly handles it.

File: tests/phpunit/tests/fonts/font-library/wpFontsDir.php
Comment (swissspidy) [inline tests/phpunit/tests/fonts/font-library/wpFontsDir.php:76]:
```suggestion
	/**
	 * @ticket 60652
	 */
	public function test_fonts_dir_filters_do_not_trigger_infinite_loop() {
```

**Review by swissspidy — APPROVED** [review]

**Comment by peterwilsoncc** [comment]
Committed [r57868](https://core.trac.wordpress.org/changeset/57868) / https://github.com/WordPress/wordpress-develop/commit/ca8d78e5f2f50c59749398b5932e246ada0c0a37

**Comment by azaozz** [comment]
This is a comment on a closed PR that was committed to WP 6.5 but for posterity's sake I think it would be good to make a tl;dr list of the points from the above reviews.

1. This code does not actually fix the infinite loop problem as specified in the corresponding Trac ticket, it only "hides" it. The (possible) error that developers may make here is to call a function from its callback. This will always result in an infinite loop, and hiding it is quite unhelpful to the developers. Then they cannot find their error easily and may even miss it and release their code with it. 

    Imagine if PHP suddenly decided to **not** throw a syntax error when somebody made a typo in their code. How bad/annoying/unhelpful would this be? :) Unfortunately the code in this PR is something like the WordPress equivalent of this.

2. This PR introduces an anti-pattern in WP: running a filter in another filter's callback. This is the main cause for the infinite loop and is a pretty bad idea for several reasons: 
    - Developers may cause an infinite loop by calling the function where the first filter runs in their callback on the second filter. In this case calling `wp_upload_dir()` from the callback on the `font_dir` filter causes an infinite loop. This type of "double callbacks" errors are not that easy to see and figure out.
    - Plugins may remove the callback that contains the `font_dir` filter breaking the core API. This is a significant defect as it makes that WP functionality unreliable.
    - Finally this code introduces yet another "private" function in the global scope in WP. This is not a huge problem but such code should be avoided in new functionality. It is not that hard to make new WP private functions not accessible by plugins. It is also considered a "bad behavior" for plugins to use or tamper with such functions, but still some do. When that happens it needlessly makes WP core and plugin development harder.

I'd also want to make is clear that I'm not trying to blame anybody whit this review. This PR was reviewed and committed during the "late" RC stage of WP 6.5, only few days before the (planned) release. Everybody was under a significant pressure at that time, and mistakes can happen at any point. The follow-up Trac ticket that fixes these issues is https://core.trac.wordpress.org/ticket/60835.

**Comment by peterwilsoncc** [comment]
@azaozz I'm not going to debate the merit of defensive coding so future maintainers of the code can perform upkeep **without the risk of breaking a filter been used in the way it is intended to be used.** It's just pointless.

This PR was approved by a tech lead after multiple reviews in the two weeks leading up to its commit.

**Comment by azaozz** [comment]
@peterwilsoncc Thanks for you replay. 

> I'm not going to debate the merit...

Sorry that I didn't make myself clear enough. I'm not discussing or debating anything here. The above comment is an attempt for a tl;dr of my review of this PR, nothing more.

> This PR was approved by a tech lead...

Right. It was approved on the condition that this code will be fixed in WP 6.6. See https://github.com/WordPress/wordpress-develop/pull/6211#issuecomment-2011793974.

> without the risk of breaking a filter been used...

Sorry but not sure what you mean by that. Are you saying that it is okay for the newly introduced `font_dir` filter to be inconsistent? Or perhaps that it is optional? How would plugins and eventually core use that filter then? 
In any case, if you want to talk about how to fix this code, please lets do that on the Trac ticket for it: https://core.trac.wordpress.org/ticket/60835.

**Pull request closed without merging**
